p2panda-store/src/sqlite/store.rs

```diff
@@ -0,0 +1,525 @@
+
+//! SQLite persistent storage.
+use std::hash::{DefaultHasher, Hash as StdHash, Hasher};
+use std::marker::PhantomData;
+
+use anyhow::{Error, Result};
+use sqlx::migrate;
+use sqlx::migrate::MigrateDatabase;
+use sqlx::sqlite::{SqlitePool, SqlitePoolOptions};
+use sqlx::{query, query_as, Sqlite};
+
+use p2panda_core::{Body, Extensions, Hash, Header, RawOperation};
+
+use crate::sqlite::models::{OperationRow, RawOperationRow};
+use crate::{LogId, OperationStore};
+
+/// Re-export of SQLite connection pool type.
+pub type Pool = SqlitePool;
+
+/// SQLite-based persistent store.
+#[derive(Clone, Debug)]
+pub struct SqliteStore<L, E> {
+    pub(crate) pool: Pool,
+    _marker: PhantomData<(L, E)>,
+}
+
+impl<L, E> SqliteStore<L, E>
+where
+    L: LogId,
+    E: Extensions,
+{
+    /// Create a new `SqliteStore` using the provided db `Pool`.
+    pub fn new(pool: Pool) -> Self {
+        Self {
+            pool,
+            _marker: PhantomData {},
+        }
+    }
+}
+
+/// Create the database if it doesn't already exist.
+pub async fn create_database(url: &str) -> Result<()> {
+    if !Sqlite::database_exists(url).await? {
+        Sqlite::create_database(url).await?;
+    }
+
+    Ok(())
+}
+
+/// Drop the database if it exists.
+pub async fn drop_database(url: &str) -> Result<()> {
+    if Sqlite::database_exists(url).await? {
+        Sqlite::drop_database(url).await?;
+    }
+
+    Ok(())
+}
+
+/// Create a connection pool.
+pub async fn connection_pool(url: &str, max_connections: u32) -> Result<Pool, Error> {
+    let pool: Pool = SqlitePoolOptions::new()
+        .max_connections(max_connections)
+        .connect(url)
+        .await?;
+
+    Ok(pool)
+}
+
+/// Run any pending database migrations from inside the application.
+pub async fn run_pending_migrations(pool: &Pool) -> Result<()> {
+    migrate!().run(pool).await?;
+    Ok(())
+}
+
+fn calculate_hash<T: StdHash>(t: &T) -> u64 {
+    let mut s = DefaultHasher::new();
+    t.hash(&mut s);
+    s.finish()
+}
+
+fn serialize_extensions<T: Extensions>(extensions: &T) -> Result<Vec<u8>> {
+    let mut bytes: Vec<u8> = Vec::new();
+    ciborium::ser::into_writer(extensions, &mut bytes)?;
+
+    Ok(bytes)
+}
+
+pub(crate) fn deserialize_extensions<T>(bytes: Vec<u8>) -> Result<T>
+where
+    T: Extensions,
+{
+    let extensions = ciborium::de::from_reader(&bytes[..])?;
+
+    Ok(extensions)
+}
+
+impl<L, E> OperationStore<L, E> for SqliteStore<L, E>
+where
+    L: LogId + Send + Sync,
+    E: Extensions + Send + Sync,
+{
+    type Error = Error;
+
+    async fn insert_operation(
+        &mut self,
+        hash: Hash,
+        header: &Header<E>,
+        body: Option<&Body>,
+        header_bytes: &[u8],
+        log_id: &L,
+    ) -> Result<bool, Self::Error> {
+        // Start a transaction.
+        //
+        // Any insertions after this point, and before `execute()`, will be rolled back in the
+        // event of an error.
+        let mut tx = self.pool.begin().await?;
```

Comment:
Is this transaction required? I understood that if only one query was being made then rollback is not an issue (as it either succeeds or fails). Maybe it's just required for the sqlx api....

Reply:
Yeh, the pool itself implements [`Executor`](https://docs.rs/sqlx/latest/sqlx/trait.Executor.html) so I believe you can pass it directly into the `execute()` method.

Reply:
Ooh yeah, nice one! Much better.